packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -6267,46 +6266,77 @@ abstract class RenderObjectElement extends Element {
-    ParentDataElement<ParentData>? result;
+    final List<ParentDataElement<ParentData>> result = <ParentDataElement<ParentData>>[];
+    final Set<Type> ancestorTypes = <Type>{};
     while (ancestor != null && ancestor is! RenderObjectElement) {
       if (ancestor is ParentDataElement<ParentData>) {
-        result = ancestor;
-        break;
+        ancestorTypes.add(ancestor.runtimeType);
+        result.add(ancestor);
       }
       ancestor = ancestor._parent;
     }
     assert(() {
-      if (result == null || ancestor == null) {
+      if (result.isEmpty || ancestor == null) {
         return true;
       }
-      // Check that no other ParentDataWidgets want to provide parent data.
-      final List<ParentDataElement<ParentData>> badAncestors = <ParentDataElement<ParentData>>[];
-      ancestor = ancestor!._parent;
-      while (ancestor != null && ancestor is! RenderObjectElement) {
-        if (ancestor is ParentDataElement<ParentData>) {
-          badAncestors.add(ancestor! as ParentDataElement<ParentData>);
-        }
-        ancestor = ancestor!._parent;
-      }
-      if (badAncestors.isNotEmpty) {
-        badAncestors.insert(0, result);
+      // Check that no other ParentDataWidgets of the same type want to provide
+      // parent data.
+      // The actual type of ParentData is checked in debugIsValidRenderObject.
+      if (ancestorTypes.length != result.length) {
+        // This can only occur if the Set of ancestors was provided a dupe and
+        // did not add it.
+        assert(ancestorTypes.length < result.length);
         try {
           // We explicitly throw here (even though we immediately redirect the
           // exception elsewhere) so that debuggers will notice it when they
           // have "break on exception" enabled.
           throw FlutterError.fromParts(<DiagnosticsNode>[
             ErrorSummary('Incorrect use of ParentDataWidget.'),
-            ErrorDescription('The following ParentDataWidgets are providing parent data to the same RenderObject:'),
-            for (final ParentDataElement<ParentData> ancestor in badAncestors)
-              ErrorDescription('- ${ancestor.widget} (typically placed directly inside a ${(ancestor.widget as ParentDataWidget<ParentData>).debugTypicalAncestorWidgetClass} widget)'),
-            ErrorDescription('However, a RenderObject can only receive parent data from at most one ParentDataWidget.'),
-            ErrorHint('Usually, this indicates that at least one of the offending ParentDataWidgets listed above is not placed directly inside a compatible ancestor widget.'),
-            ErrorDescription('The ownership chain for the RenderObject that received the parent data was:\n  ${debugGetCreatorChain(10)}'),
+            ErrorDescription(
+              'Competing ParentDataWidgets are providing parent data to the '
+              'same RenderObject:'
+            ),
+            for (final ParentDataElement<ParentData> ancestor in result)
```

Comment:
I wonder if we can improve this exception by only listing the conflicting ParentDataWidgets, e.g. if you are trying to apply two `Foo` ParentDataWidgets and one `Bar` ParentDataWidget, should we only list the two `Foo`s since the `Bar` widget is kinda irrelevant to the problem.

(Or we could at least list them in two different sections in the exception for extra context: Here are the conflicting ones, and here are all other ones that try to contribute data. However, I think you really only care about the conflicting ones...)

Reply:
(if we do this, we should add an extra test for it)

Reply:
I ended up doing this. :+1: